File: src/api/tools.ts
import { invoke } from "@tauri-apps/api/core";

import type { PersonaToolDefinition } from "@/lib/bindings/PersonaToolDefinition";
import type { CreateToolDefinitionInput } from "@/lib/bindings/CreateToolDefinitionInput";
import type { UpdateToolDefinitionInput } from "@/lib/bindings/UpdateToolDefinitionInput";
import type { PersonaTool } from "@/lib/bindings/PersonaTool";
import type { ToolUsageSummary, ToolUsageOverTime, PersonaUsageSummary } from "@/lib/types/types";

// ============================================================================
// Tool Definitions & Assignments
// ============================================================================

export const listToolDefinitions = () =>
  invoke<PersonaToolDefinition[]>("list_tool_definitions");

export const getToolDefinition = (id: string) =>
  invoke<PersonaToolDefinition>("get_tool_definition", { id });

export const getToolDefinitionsByCategory = (category: string) =>
  invoke<PersonaToolDefinition[]>("get_tool_definitions_by_category", { category });

export const createToolDefinition = (input: CreateToolDefinitionInput) =>
  invoke<PersonaToolDefinition>("create_tool_definition", { input });

export const updateToolDefinition = (id: string, input: UpdateToolDefinitionInput) =>
  invoke<PersonaToolDefinition>("update_tool_definition", { id, input });

export const deleteToolDefinition = (id: string) =>
  invoke<boolean>("delete_tool_definition", { id });

export const assignTool = (
  personaId: string,
  toolId: string,
  toolConfig?: string,
) =>
  invoke<PersonaTool>("assign_tool", {
    personaId,
    toolId,
    toolConfig: toolConfig ?? null,
  });

export const unassignTool = (personaId: string, toolId: string) =>
  invoke<boolean>("unassign_tool", { personaId, toolId });

export const bulkAssignTools = (personaId: string, toolIds: string[]) =>
  invoke<number>("bulk_assign_tools", { personaId, toolIds });

export const bulkUnassignTools = (personaId: string, toolIds: string[]) =>
  invoke<number>("bulk_unassign_tools", { personaId, toolIds });

// ── Tool Usage Analytics ──────────────────────────────────────────────────

export const getToolUsageSummary = (since: string, personaId?: string) =>
  invoke<ToolUsageSummary[]>("get_tool_usage_summary", { since, personaId: personaId ?? null });

export const getToolUsageOverTime = (since: string, personaId?: string) =>
  invoke<ToolUsageOverTime[]>("get_tool_usage_over_time", { since, personaId: personaId ?? null });

export const getToolUsageByPersona = (since: string) =>
  invoke<PersonaUsageSummary[]>("get_tool_usage_by_persona", { since });

File: src-tauri/src/commands/tools/tools.rs
use std::sync::Arc;
use tauri::State;

use crate::db::models::{CreateToolDefinitionInput, PersonaTool, PersonaToolDefinition, UpdateToolDefinitionInput};
use crate::db::repos::execution::tool_usage;
use crate::db::repos::resources::tools as repo;
use crate::error::AppError;
use crate::AppState;

#[tauri::command]
pub fn list_tool_definitions(
    state: State<'_, Arc<AppState>>,
) -> Result<Vec<PersonaToolDefinition>, AppError> {
    repo::get_all_definitions(&state.db)
}

#[tauri::command]
pub fn get_tool_definition(
    state: State<'_, Arc<AppState>>,
    id: String,
) -> Result<PersonaToolDefinition, AppError> {
    repo::get_definition_by_id(&state.db, &id)
}

#[tauri::command]
pub fn get_tool_definitions_by_category(
    state: State<'_, Arc<AppState>>,
    category: String,
) -> Result<Vec<PersonaToolDefinition>, AppError> {
    repo::get_definitions_by_category(&state.db, &category)
}

#[tauri::command]
pub fn create_tool_definition(
    state: State<'_, Arc<AppState>>,
    input: CreateToolDefinitionInput,
) -> Result<PersonaToolDefinition, AppError> {
    repo::create_definition(&state.db, input)
}

#[tauri::command]
pub fn update_tool_definition(
    state: State<'_, Arc<AppState>>,
    id: String,
    input: UpdateToolDefinitionInput,
) -> Result<PersonaToolDefinition, AppError> {
    repo::update_definition(&state.db, &id, input)
}

#[tauri::command]
pub fn delete_tool_definition(
    state: State<'_, Arc<AppState>>,
    id: String,
) -> Result<bool, AppError> {
    repo::delete_definition(&state.db, &id)
}

#[tauri::command]
pub fn assign_tool(
    state: State<'_, Arc<AppState>>,
    persona_id: String,
    tool_id: String,
    tool_config: Option<String>,
) -> Result<PersonaTool, AppError> {
    repo::assign_tool(&state.db, &persona_id, &tool_id, tool_config)
}

#[tauri::command]
pub fn unassign_tool(
    state: State<'_, Arc<AppState>>,
    persona_id: String,
    tool_id: String,
) -> Result<bool, AppError> {
    repo::unassign_tool(&state.db, &persona_id, &tool_id)
}

#[tauri::command]
pub fn bulk_assign_tools(
    state: State<'_, Arc<AppState>>,
    persona_id: String,
    tool_ids: Vec<String>,
) -> Result<u32, AppError> {
    repo::bulk_assign_tools(&state.db, &persona_id, &tool_ids)
}

#[tauri::command]
pub fn bulk_unassign_tools(
    state: State<'_, Arc<AppState>>,
    persona_id: String,
    tool_ids: Vec<String>,
) -> Result<u32, AppError> {
    repo::bulk_unassign_tools(&state.db, &persona_id, &tool_ids)
}

#[tauri::command]
pub fn get_tool_usage_summary(
    state: State<'_, Arc<AppState>>,
    since: String,
    persona_id: Option<String>,
) -> Result<Vec<serde_json::Value>, AppError> {
    tool_usage::get_usage_summary(&state.db, &since, persona_id.as_deref())
}

#[tauri::command]
pub fn get_tool_usage_over_time(
    state: State<'_, Arc<AppState>>,
    since: String,
    persona_id: Option<String>,
) -> Result<Vec<serde_json::Value>, AppError> {
    tool_usage::get_usage_over_time(&state.db, &since, persona_id.as_deref())
}

#[tauri::command]
pub fn get_tool_usage_by_persona(
    state: State<'_, Arc<AppState>>,
    since: String,
) -> Result<Vec<serde_json::Value>, AppError> {
    tool_usage::get_usage_by_persona(&state.db, &since)
}
